core/state/statedb.go

```diff
@@ -1373,64 +1367,91 @@ func (s *StateDB) Commit(block uint64, deleteEmptyObjects bool) (common.Hash, er
 // - Reset transient storage (EIP-1153)
 func (s *StateDB) Prepare(rules params.Rules, sender, coinbase common.Address, dst *common.Address, precompiles []common.Address, list types.AccessList) {
 	if rules.IsBerlin {
-		// Clear out any leftover from previous executions
-		al := newAccessList()
-		s.accessList = al
+		if !rules.IsPrague {
+			s.accessList = newAccessList()
+		} else {
+			s.accessList = s.NewAccessWitness()
+		}
+		al := s.accessList
 
-		al.AddAddress(sender)
+		al.AddAddress(sender, ALAllItems, false)
 		if dst != nil {
-			al.AddAddress(*dst)
+			al.AddAddress(*dst, ALAllItems, false)
 			// If it's a create-tx, the destination will be added inside evm.create
 		}
 		for _, addr := range precompiles {
-			al.AddAddress(addr)
+			al.AddAddress(addr, 0, false)
 		}
 		for _, el := range list {
-			al.AddAddress(el.Address)
+			al.AddAddress(el.Address, 0, false)
 			for _, key := range el.StorageKeys {
-				al.AddSlot(el.Address, key)
+				al.AddSlot(el.Address, key, false)
 			}
 		}
 		if rules.IsShanghai { // EIP-3651: warm coinbase
-			al.AddAddress(coinbase)
+			al.AddAddress(coinbase, ALAllItems, false)
 		}
 	}
 	// Reset transient storage at the beginning of transaction execution
 	s.transientStorage = newTransientStorage()
 }
 
 // AddAddressToAccessList adds the given address to the access list
-func (s *StateDB) AddAddressToAccessList(addr common.Address) {
-	if s.accessList.AddAddress(addr) {
+func (s *StateDB) AddAddressToAccessList(addr common.Address, mode ALAccountItem, isWrite ALAccessMode) uint64 {
+	gas := s.accessList.AddAddress(addr, mode, isWrite)
+	if gas > 0 {
 		s.journal.append(accessListAddAccountChange{&addr})
 	}
+	return gas
 }
 
 // AddSlotToAccessList adds the given (address, slot)-tuple to the access list
-func (s *StateDB) AddSlotToAccessList(addr common.Address, slot common.Hash) {
-	addrMod, slotMod := s.accessList.AddSlot(addr, slot)
-	if addrMod {
-		// In practice, this should not happen, since there is no way to enter the
-		// scope of 'address' without having the 'address' become already added
-		// to the access list (via call-variant, create, etc).
-		// Better safe than sorry, though
-		s.journal.append(accessListAddAccountChange{&addr})
-	}
-	if slotMod {
+func (s *StateDB) AddSlotToAccessList(addr common.Address, slot common.Hash, isWrite ALAccessMode) uint64 {
+	gas := s.accessList.AddSlot(addr, slot, isWrite)
+	if gas > params.WarmStorageReadCostEIP2929 {
```

Comment:
s `WarmStorageReadCostEIP2929` the minimum non-zero value that can be returned considering the two potential implementations of access list? Any reason not to use `>0` that might be clearer?

Reply:
I actually think `params.WarmStorageReadCostEIP2929`, which is indeed the smallest value that can be returned at present, is clearer because you know that this is the minimum to be expected.